File: src/text/tag.rs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

//! Tag normalization and the sub-path match predicate (ADRs 0006, 0023).
//!
//! A tag is a `/`-separated hierarchy. Each segment is normalized with the slug
//! rules, the `/` separators are preserved, and empty segments are dropped so a
//! tag canonicalizes to a clean lowercase form (`Rust` and `rust` are the same
//! tag). Matching a query tag against a note tag is a contiguous-segment-run
//! search, which is the semantics the `tag:` query predicate evaluates.

use super::slug::normalize_segment;

/// Split a tag into its normalized, non-empty segments.
///
/// Each `/`-delimited piece is run through the slug segment normalizer; pieces
/// that normalize to nothing (e.g. from `a//b` or a stray separator) are
/// dropped. This is the canonical decomposition every other tag operation is
/// built on.
pub fn segments(tag: &str) -> Vec<String> {
    tag.split('/')
        .map(normalize_segment)
        .filter(|seg| !seg.is_empty())
        .collect()
}

/// Normalize a tag to its canonical `a/b/c` string form.
///
/// Returns an empty string when the tag has no surviving segments, which the
/// note model treats as "no tag".
pub fn normalize(tag: &str) -> String {
    segments(tag).join("/")
}

/// Test whether `query` matches `candidate` under the sub-path rule (ADR 0006
/// as refined for v1).
///
/// Both sides are decomposed into normalized segments; the query matches iff
/// its segment list occurs as a contiguous run of full segments anywhere in the
/// candidate's segment list. So `programming` matches `programming`,
/// `programming/foo`, `bar/programming` and `baz/programming/blub`, while
/// `programming/foo` matches any tag containing that consecutive chain.
/// Normalization makes the match case-insensitive. An empty query matches
/// nothing.
pub fn matches(query: &str, candidate: &str) -> bool {
    let needle = segments(query);
    if needle.is_empty() {
        return false;
    }
    let haystack = segments(candidate);
    haystack
        .windows(needle.len())
        .any(|window| window == needle)
}

/// Existing tags that could complete the partial tag being typed.
///
/// Matching is hierarchy-aware. The partial's leading segments must equal the
/// candidate's, and the candidate's segment at the partial's depth must start
/// with the partial's last segment, so `prog` offers `programming/rust` and
/// `programming/ru` narrows to `programming/rust`. A partial ending in `/`
/// offers the strictly deeper children below it. An empty partial offers every
/// candidate. The partial is normalized, so matching is case-insensitive.
/// Order is preserved and duplicates are dropped.
pub fn suggest(partial: &str, candidates: &[String]) -> Vec<String> {
    let needle = segments(partial);
    let trailing_slash = partial.ends_with('/');

    let mut seen = std::collections::HashSet::new();
    candidates
        .iter()
        .filter(|candidate| {
            if needle.is_empty() {
                return true;
            }
            let haystack = segments(candidate);
            if trailing_slash {
                // Offer children strictly below the completed partial.
                haystack.len() > needle.len() && haystack[..needle.len()] == needle[..]
            } else {
                let (last, parents) = needle.split_last().expect("needle is non-empty");
                haystack.len() >= needle.len()
                    && haystack[..parents.len()] == parents[..]
                    && haystack[parents.len()].starts_with(last)
            }
        })
        .filter(|candidate| seen.insert((*candidate).clone()))
        .cloned()
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn tags(values: &[&str]) -> Vec<String> {
        values.iter().map(|v| (*v).to_owned()).collect()
    }

    #[test]
    fn suggest_empty_partial_offers_all() {
        let all = tags(&["programming/rust", "area/work"]);
        assert_eq!(suggest("", &all), all);
    }

    #[test]
    fn suggest_prefixes_the_first_segment() {
        let all = tags(&["programming/rust", "programming/cli", "area/work"]);
        assert_eq!(
            suggest("prog", &all),
            tags(&["programming/rust", "programming/cli"])
        );
    }

    #[test]
    fn suggest_trailing_slash_offers_children() {
        let all = tags(&["programming", "programming/rust", "programming/cli"]);
        assert_eq!(
            suggest("programming/", &all),
            tags(&["programming/rust", "programming/cli"])
        );
    }

    #[test]
    fn suggest_narrows_within_a_level() {
        let all = tags(&["programming/rust", "programming/cli"]);
        assert_eq!(suggest("programming/ru", &all), tags(&["programming/rust"]));
    }

    #[test]
    fn suggest_is_case_insensitive_and_normalizing() {
        let all = tags(&["ueber/groesse"]);
        assert_eq!(suggest("Über", &all), tags(&["ueber/groesse"]));
    }

    #[test]
    fn suggest_drops_duplicates_and_non_matches() {
        let all = tags(&["area/work", "area/work", "programming/rust"]);
        assert_eq!(suggest("area", &all), tags(&["area/work"]));
        assert!(suggest("xyz", &all).is_empty());
    }

    #[test]
    fn segments_normalize_each_piece() {
        assert_eq!(segments("Programming/Rust"), vec!["programming", "rust"]);
        assert_eq!(segments("Area/Work"), vec!["area", "work"]);
    }

    #[test]
    fn segments_drop_empty_pieces() {
        assert_eq!(segments("a//b"), vec!["a", "b"]);
        assert_eq!(segments("/leading/"), vec!["leading"]);
        assert_eq!(segments(""), Vec::<String>::new());
    }

    #[test]
    fn normalize_roundtrips_to_canonical_string() {
        assert_eq!(normalize("Programming/Rust"), "programming/rust");
        assert_eq!(normalize("Über/Größe"), "ueber/groesse");
        assert_eq!(normalize("///"), "");
    }

    #[test]
    fn matches_single_segment_anywhere() {
        assert!(matches("programming", "programming"));
        assert!(matches("programming", "programming/foo"));
        assert!(matches("programming", "bar/programming"));
        assert!(matches("programming", "baz/programming/blub"));
    }

    #[test]
    fn matches_multi_segment_contiguous_run() {
        assert!(matches("programming/foo", "programming/foo"));
        assert!(matches("programming/foo", "x/programming/foo/y"));
        // Not contiguous: a segment intervenes.
        assert!(!matches("programming/foo", "programming/bar/foo"));
    }

    #[test]
    fn matches_is_case_insensitive() {
        assert!(matches("Rust", "programming/rust"));
        assert!(matches("rust", "Programming/Rust"));
    }

    #[test]
    fn matches_requires_full_segment_not_substring() {
        // `prog` is not a segment of `programming`.
        assert!(!matches("prog", "programming"));
    }

    #[test]
    fn empty_query_matches_nothing() {
        assert!(!matches("", "anything"));
        assert!(!matches("///", "anything"));
    }
}

File: site/src/search/tag.ts
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

// Tag normalization and the sub-path match, mirroring `src/text/slug.rs` and
// `src/text/tag.rs` (ADR 0023, ADR 0006). The shared conformance corpus pins
// the two implementations against each other (ADR 0052).

const MAX_SEGMENT_LEN = 72;

const TRANSLITERATIONS: Record<string, string> = {
  ä: "ae",
  ö: "oe",
  ü: "ue",
  ß: "ss",
  Ä: "Ae",
  Ö: "Oe",
  Ü: "Ue",
  à: "a",
  á: "a",
  â: "a",
  ã: "a",
  å: "a",
  ā: "a",
  ª: "a",
  À: "A",
  Á: "A",
  Â: "A",
  Ã: "A",
  Å: "A",
  Ā: "A",
  è: "e",
  é: "e",
  ê: "e",
  ë: "e",
  ē: "e",
  ė: "e",
  ę: "e",
  È: "E",
  É: "E",
  Ê: "E",
  Ë: "E",
  Ē: "E",
  ì: "i",
  í: "i",
  î: "i",
  ï: "i",
  ī: "i",
  į: "i",
  Ì: "I",
  Í: "I",
  Î: "I",
  Ï: "I",
  Ī: "I",
  ò: "o",
  ó: "o",
  ô: "o",
  õ: "o",
  ø: "o",
  ō: "o",
  Ò: "O",
  Ó: "O",
  Ô: "O",
  Õ: "O",
  Ø: "O",
  Ō: "O",
  ù: "u",
  ú: "u",
  û: "u",
  ū: "u",
  ů: "u",
  Ù: "U",
  Ú: "U",
  Û: "U",
  Ū: "U",
  ñ: "n",
  Ñ: "N",
  ç: "c",
  Ç: "C",
  ý: "y",
  ÿ: "y",
};

/** One character's ASCII transliteration; other non-ASCII drops out. */
function transliterate(ch: string): string {
  const mapped = TRANSLITERATIONS[ch];
  if (mapped !== undefined) return mapped;
  return ch.charCodeAt(0) < 0x80 ? ch : "";
}

/**
 * The slug normalization pipeline without the `untitled` fallback:
 * transliterate, lowercase, whitespace runs to `-`, keep `[a-z0-9-]`,
 * collapse and trim dashes, cap the length at a dash boundary.
 */
export function normalizeSegment(input: string): string {
  const transliterated = Array.from(input).map(transliterate).join("");
  const lowered = transliterated.toLowerCase();

  let spaced = "";
  let prevSpace = false;
  for (const ch of lowered) {
    if (/\s/.test(ch)) {
      if (!prevSpace) spaced += "-";
      prevSpace = true;
    } else {
      spaced += ch;
      prevSpace = false;
    }
  }

  const filtered = spaced.replace(/[^a-z0-9-]/g, "");
  const collapsed = filtered.replace(/-+/g, "-").replace(/^-|-$/g, "");
  return truncateAtBoundary(collapsed, MAX_SEGMENT_LEN);
}

function truncateAtBoundary(input: string, max: number): string {
  if (input.length <= max) return input;
  const window = input.slice(0, max);
  const cut = window.lastIndexOf("-");
  const kept = cut === -1 ? window : window.slice(0, cut);
  return kept.replace(/^-+|-+$/g, "");
}

/** A tag's normalized, non-empty segments. */
export function segments(tag: string): string[] {
  return tag
    .split("/")
    .map(normalizeSegment)
    .filter((segment) => segment.length > 0);
}

/** A tag's canonical `a/b/c` form; empty when nothing survives. */
export function normalizeTag(tag: string): string {
  return segments(tag).join("/");
}

/**
 * The sub-path rule: the query's segments occur as a contiguous run of full
 * segments anywhere in the candidate's. An empty query matches nothing.
 */
export function tagMatches(query: string, candidate: string): boolean {
  const needle = segments(query);
  if (needle.length === 0) return false;
  const haystack = segments(candidate);
  for (let start = 0; start + needle.length <= haystack.length; start++) {
    let same = true;
    for (let i = 0; i < needle.length; i++) {
      if (haystack[start + i] !== needle[i]) {
        same = false;
        break;
      }
    }
    if (same) return true;
  }
  return false;
}

/**
 * The tags a note carries, as the note parser lifts them from frontmatter:
 * a list or a single scalar, each normalized, empties dropped, duplicates
 * removed in order of first appearance.
 */
export function tagsOf(raw: unknown): string[] {
  const items = Array.isArray(raw)
    ? raw
    : raw === null || raw === undefined
      ? []
      : [raw];
  const tags: string[] = [];
  for (const item of items) {
    if (
      typeof item !== "string" &&
      typeof item !== "number" &&
      typeof item !== "boolean"
    ) {
      continue;
    }
    const tag = normalizeTag(String(item));
    if (tag.length > 0 && !tags.includes(tag)) tags.push(tag);
  }
  return tags;
}
